src/Authentication/Rps/RpsAuthenticator.cs

```diff
@@ -0,0 +1,140 @@
+{
+    /// <summary>
+    /// Handler and token logic for RPS (Live ID) authentication.
+    /// </summary>
+    /// <seealso cref="IAuthenticator" />
+    public class RpsAuthenticator : IAuthenticator
+    {
+        private readonly IRestfulClient tokenClient;
+        private readonly ReaderWriterLockSlim lockSlim;
+        private readonly IMemoryCache cache;
+
+        /// <summary>
+        /// Gets or sets the configuration for RPS (Live ID) requests.
+        /// </summary>
+        public IRpsConfig Config { get; protected set; }
+
+        /// <summary>
+        /// Gets the query string used in RPS (Live ID) requests.
+        /// </summary>
+        public string QueryString => $"grant_type=client_credentials&client_id={Config.SiteId}&scope={ServiceScope}";
+
+        /// <summary>
+        /// Gets the service scope used in RPS (Live ID) requests.
+        /// </summary>
+        public string ServiceScope => $"{Config.Scope}::{Config.Policy}";
+
+        private string TokenCacheKey => $"rpsToken::{ServiceScope}";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RpsAuthenticator"/> class.
+        /// Base constructor.
+        /// </summary>
+        /// <param name="config">Configuration for RPS (Live ID) configuration.</param>
+        public RpsAuthenticator(IRpsConfig config)
+            : this(config, null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RpsAuthenticator"/> class.
+        /// </summary>
+        /// <param name="config">Configuration for RPS (Live ID) authentication.</param>
+        /// <param name="clientHandler">Client handler for RPS auth requests.</param>
+        /// <param name="delegatingHandlers">Delegating handlers for RPS auth requests.</param>
+        public RpsAuthenticator(IRpsConfig config, HttpClientHandler clientHandler, params DelegatingHandler[] delegatingHandlers)
+        {
+            Config = config;
+
+            tokenClient = new RestfulClient(config.AuthUrl, EnhanceClientHandler(clientHandler), delegatingHandlers);
+            lockSlim = new ReaderWriterLockSlim();
+            cache = new MemoryCache(new MemoryCacheOptions());
+        }
+
+        /// <inheritdoc/>
+        public HttpClientHandler EnhanceClientHandler(HttpClientHandler clientHandler)
+        {
+            var handler = clientHandler ?? new HttpClientHandler();
+
+            handler.SslProtocols = SslProtocols.Tls12;
+            handler.ClientCertificates.Add(Config.SiteCertificate);
+
+            return handler;
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> GetAuthValue()
+        {
+            RpsToken token;
+
+            lockSlim.EnterUpgradeableReadLock();
+            try
+            {
+                token = cache.Get(TokenCacheKey) as RpsToken;
+
+                if (token == null)
+                {
+                    lockSlim.EnterWriteLock();
```

Comment:
Aad's authentication context handles this under the covers. If another auth provider integration needs this locking, I can pull it out.